From c86b552689002ca89727cdbddeae4c15bfd85f1c Mon Sep 17 00:00:00 2001
From: Heinrich Toews <heinrich.toews@wago.com>
Date: Thu, 14 May 2020 11:50:17 +0200
Subject: [PATCH] defaultenv: defaultenv-pfc: move switch reset to non-default
 boot mode

For a separated mode switch setup it is essential to keep the switch ports down
until the switch configuraiton is done. We remove the switch reset during
normal boot, for the linux kernel is doing that for us.

pfc200-820x devices have a different HW setup which gets the switch
out of reset automatically. In this case we need to power down the ports as
soon as possible and we do that during default boot.

Signed-off-by: Heinrich Toews <heinrich.toews@wago.com>
---
 .../init/pfc-ksz8863-pwr-down                      | 26 ++++++++++++++++++++++
 .../defaultenv-pfc-base/init/a-pfc-switch-reset    | 18 ---------------
 .../init_post_autoboot/a-pfc-switch-reset          | 22 ++++++++++++++++++
 3 files changed, 48 insertions(+), 18 deletions(-)
 create mode 100644 arch/arm/boards/wago-pfc-am35xx/defaultenv-pfc-am35xx/init/pfc-ksz8863-pwr-down
 delete mode 100644 defaultenv/defaultenv-pfc-base/init/a-pfc-switch-reset
 create mode 100644 defaultenv/defaultenv-pfc-base/init_post_autoboot/a-pfc-switch-reset

diff --git a/arch/arm/boards/wago-pfc-am35xx/defaultenv-pfc-am35xx/init/pfc-ksz8863-pwr-down b/arch/arm/boards/wago-pfc-am35xx/defaultenv-pfc-am35xx/init/pfc-ksz8863-pwr-down
new file mode 100644
index 0000000..d038b38
--- /dev/null
+++ b/arch/arm/boards/wago-pfc-am35xx/defaultenv-pfc-am35xx/init/pfc-ksz8863-pwr-down
@@ -0,0 +1,26 @@
+#!/bin/sh
+
+source env/data/print-msg-formats
+
+# mii bus
+mii_bus=0
+
+# mii addresses
+phy_x1_addr=1
+phy_x2_addr=2
+
+# registers
+control_reg=0
+
+powerdown_enable=6176    	# 0x1820 - set the power down flag
+
+# We assume that we are running on a pfc200 G1 device
+#
+# Power down the phys. This is essetial especially for separated mode
+# operation.
+
+echo -e ${info} "switch: power down the phys"
+miitool -w "${mii_bus}:${phy_x1_addr}:${control_reg}" -p powerdown_enable
+miitool -w "${mii_bus}:${phy_x2_addr}:${control_reg}" -p powerdown_enable
+
+exit 0
diff --git a/defaultenv/defaultenv-pfc-base/init/a-pfc-switch-reset b/defaultenv/defaultenv-pfc-base/init/a-pfc-switch-reset
deleted file mode 100644
index ab47598..0000000
--- a/defaultenv/defaultenv-pfc-base/init/a-pfc-switch-reset
+++ /dev/null
@@ -1,18 +0,0 @@
-#!/bin/sh
-
-source env/data/print-msg-formats
-
-# Hint: on am35xx based devices we do not need to get the switch out of reset
-# manually. This is already done by the hardware reset-logic.
-
-if [ ${global.pfc.sw_reset_gpio} -eq -1 ]; then
-	echo -e ${warning} "no switch reset pin defined for board \e[00;32m${global.board.variant}\e[00m"
-	exit 0
-else
-	echo -n -e ${info} "reset switch for board \e[00;32m${global.board.variant}\e[00m with gpio pin "
-	echo "${global.pfc.sw_reset_gpio}"
-fi
-
-gpio_direction_output "${global.pfc.sw_reset_gpio}" 0
-msleep 1
-gpio_direction_output "${global.pfc.sw_reset_gpio}" 1
diff --git a/defaultenv/defaultenv-pfc-base/init_post_autoboot/a-pfc-switch-reset b/defaultenv/defaultenv-pfc-base/init_post_autoboot/a-pfc-switch-reset
new file mode 100644
index 0000000..f2d35d5
--- /dev/null
+++ b/defaultenv/defaultenv-pfc-base/init_post_autoboot/a-pfc-switch-reset
@@ -0,0 +1,22 @@
+#!/bin/sh
+
+source env/data/print-msg-formats
+
+# Hint: on am35xx based devices we do not need to get the switch out of reset
+# manually. This is already done by the hardware reset-logic.
+
+if [ "${global.autoboot}" = 0 -a "${global.pfc.boot_mode}" = "default" ]; then
+	exit 0
+fi
+
+if [ ${global.pfc.sw_reset_gpio} -eq -1 ]; then
+	echo -e ${warning} "no switch reset pin defined for board \e[00;32m${global.board.variant}\e[00m"
+	exit 0
+else
+	echo -n -e ${info} "reset switch for board \e[00;32m${global.board.variant}\e[00m with gpio pin "
+	echo "${global.pfc.sw_reset_gpio}"
+fi
+
+gpio_direction_output "${global.pfc.sw_reset_gpio}" 0
+msleep 1
+gpio_direction_output "${global.pfc.sw_reset_gpio}" 1
-- 
2.7.4

